linalg/petsc.cpp

```diff
@@ -1767,19 +1821,11 @@ void PetscParMatrix::SetUpForDevice()
             {
                bool expT = false;
                Mat sA = sub[i][j];
-               ierr = PetscObjectTypeCompare((PetscObject)sA,MATSEQAIJ,&isseqaij);
+               ierr = PetscObjectTypeCompareAny((PetscObject)sA,&isaij,MATSEQAIJ,MATMPIAIJ,"");
                PCHKERRQ(sA,ierr);
-               ierr = PetscObjectTypeCompare((PetscObject)sA,MATMPIAIJ,&ismpiaij);
-               PCHKERRQ(sA,ierr);
-               if (isseqaij)
-               {
-                  ierr = MatSetType(sA,MATSEQAIJCUSPARSE); PCHKERRQ(sA,ierr);
-                  dvec = true;
-                  expT = true;
-               }
-               else if (ismpiaij)
+               if (isaij)
                {
-                  ierr = MatSetType(sA,MATMPIAIJCUSPARSE); PCHKERRQ(sA,ierr);
+                  ierr = MatSetType(sA,PETSC_MATAIJDEVICE); PCHKERRQ(sA,ierr);
```

Comment:
There is sometimes a preference for using `aijkokkos`. For example, `vpbjacobi` setup is done on-device. Would that be hard to handle?

I see there is no `MatSetValuesCOO` in the MFEM interface yet. Among other things, that interface can assemble directly into hypre `ParCSR` end-to-end on the GPU. Is the strategy here going to make it harder to support that in the future?

Reply:
The new device conversion also does GPU to GPU mem copies https://gitlab.com/petsc/petsc/-/merge_requests/6651/diffs?commit_id=36fc4dba622d6b0de176f88c715da72ce91949c4

My interface to MFEM predates the design of MatSetValuesCOO, which I have introduced for OpenFOAM. I'm slowly transitioning the HYPRE code in PETSc to use COO inside conversion operations.

Reply:
I have full HYPRE<->AIJ GPUtoGPU conversions in the PETSc MR now